File: Python/find_numbers_with_even_number_of_digits.py
from ast import List


class Solution:
    """
    # Approach 1: Divide by 10 until number becomes 0 and count digits
    # Time Complexity: O(n *log(max num of digits)) where n is array length and d is average digits per number
    # Space Complexity: O(1)
    # Logic: For each number, repeatedly divide by 10 to count digits.
    #        If digit count is even, increment the counter.
    """
    def numberHasEvenDigits(self, num:int)->bool:
        digitcount = 0
        while num!=0:
            num = num//10
            digitcount +=1

        return digitcount%2==0 # True if even, False if odd

    def findNumbers(self, nums: List[int]) -> int:
        even_count = 0
        for num in nums:
            if(self.numberHasEvenDigits(num)):
                even_count+=1

        return even_count
class Solution2:
    """
    # Approach 2: Convert number to string and check length
    # Time Complexity: O(n * d) where n is array length and d is average digits per number
    # Space Complexity: O(1)
    # Logic: Convert each number to string and check if length is even.
    """
    def findNumbers(self, nums: List[int]) -> int:
        even_count = 0
        for num in nums:
            if len(str(num)) % 2 == 0:
                even_count += 1

        return even_count
if __name__ == "__main__":
    solution = Solution()
    
    # Sample test cases
    test_cases = [
        [12, 345, 2, 6, 7896],
        [555, 901, 482, 1771],
        [1, 22, 333, 4444],
        [1234, 56789, 0, 12]
    ]
    
    for i, nums in enumerate(test_cases, 1):
        result = solution.findNumbers(nums)
        print(f"Test case {i}: {nums}")
        print(f"Output: {result}")
        print()

File: Java/find_numbers_with_even_number_of_digits.java
class Solution {
    //first approach nlogn
    public boolean numHasEvenDigit(int num){
        int digitCount = 0;
        while(num!=0){
            num = num/10;
            digitCount++;
        }
        return digitCount%2 ==0;
    }
    public int findNumbers(int[] nums) {
        int even_count = 0;
        for(int i=0;i<nums.length;i++){
            if(numHasEvenDigit(nums[i])){
                even_count++;
            }

        } 
        return even_count;
    }
}
public class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
    
    int[][] testCases = {
        {12,345,2,6,7896},
        {555,901,482,1771},
        {1,22,333,4444,55555}
    };
    
    for (int i = 0; i < testCases.length; i++) {
        int result = solution.findNumbers(testCases[i]);
        System.out.print("Test case " + (i + 1) + ": [");
        for (int j = 0; j < testCases[i].length; j++) {
            System.out.print(testCases[i][j]);
            if (j < testCases[i].length - 1) System.out.print(", ");
        }
        System.out.println("]");
        System.out.println("Output: " + result);
        System.out.println();
    }
}